test/Concurrency/isolated_default_arguments.swift

```diff
@@ -0,0 +1,182 @@
+}
+
+func nonisolatedCaller() {
+  // expected-error@+1 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  mainActorDefaultArg()
+
+  // FIXME: confusing error message.
+  // expected-error@+1 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  mainActorClosure()
+
+  // expected-error@+1 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  mainActorClosureCall()
+}
+
+func nonisolatedAsyncCaller() async {
+  // expected-error@+1 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  mainActorDefaultArg()
+
+  // expected-error@+1 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  mainActorClosure()
+
+  // expected-error@+1 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  mainActorClosureCall()
+
+  await mainActorDefaultArg(value: requiresMainActor())
+
+  // 'await' doesn't help closures in default arguments; the calling context needs a matching
+  // actor isolation for that isolation to be inferred for the closure value.
+
+  // expected-error@+2 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  // expected-warning@+1 {{no 'async' operations occur within 'await' expression}}
+  await mainActorClosure()
+
+  // expected-error@+2 {{main actor-isolated default argument cannot be synchronously evaluated from a nonisolated context}}
+  // expected-warning@+1 {{no 'async' operations occur within 'await' expression}}
+  await mainActorClosureCall()
+}
+
+func conflictingIsolationDefaultArg(
+  // expected-error@+1 {{default argument cannot be both main actor-isolated and global actor 'SomeGlobalActor'-isolated}}
+  value: (Int, Int) = (requiresMainActor(), requiresSomeGlobalActor())
+) {}
+
+func closureLosesIsolation(
+  closure: @Sendable () -> Void = {
+    // expected-note@+1 {{calls to local function 'f()' from outside of its actor context are implicitly asynchronous}}
+    @MainActor func f() {}
+    // expected-error@+1 {{call to main actor-isolated local function 'f()' in a synchronous nonisolated context}}
+    f()
+  }
+) {}
+
+func closureIsolationMismatch(
+  closure: @SomeGlobalActor () -> Void = {
+    // expected-note@+1 {{calls to local function 'f()' from outside of its actor context are implicitly asynchronous}}
+    @MainActor func f() {}
+    // expected-error@+1 {{call to main actor-isolated local function 'f()' in a synchronous global actor 'SomeGlobalActor'-isolated context}}
+    f()
+  }
+) {}
+
+func conflictingClosureIsolation(
+  // expected-error@+1 {{default argument cannot be both main actor-isolated and global actor 'SomeGlobalActor'-isolated}}
+  closure: () -> Void = {
+    _ = requiresMainActor()
+    _ = requiresSomeGlobalActor()
```

Comment:
that's a pretty nice diagnostic for these, how it falls out of the inference here... nice